### PR TITLE
ENH: stats: override stats.gumbel_r.fit and stats.gumbel_l.fit with system of equations for MLE

### PR DESCRIPTION
**Pull request by swallan**
#### Reference issue
 from the gh-11782 main issue

#### What does this implement/fix?

Implements MLE system of equations equations for gumbel_r distribution according to the equation found in the text Evans, Hastings, and Peacock (2000), Statistical Distributions, 3rd. Ed., John Wiley and Sons.pdf, page 101.

#### Additional information

A runtime warning is sometimes raised from within optimize.root because the guess causes an overflow from system of equations.

### CODE REVIEW

**Comment by swallan** [comment]
https://github.com/scipy/scipy/pull/12097#issuecomment-653731803

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:923]:
```suggestion
```
No need for this.

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:936]: I think it's fine that we're not testing for e.g. invalid values of fixed parameters because all of that is checked by the superclass `fit` method. I think the only test that should be required specific to this override of `fit` is the one you have above. ✓

File: scipy/stats/_continuous_distns.py
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:2991]: (if data is expected to be small this probably doesn't matter and can be ignored)

I'm not super familiar with how python optimizes but It seems like this would be faster and maybe easier to parse?

```
exp_x_b = np.exp(-data / b)  # avoid recomputing 2 times
sum_exp_x_b = np.sum(exp_x_b) # avoids 1 time recomputing 

x1 = data.mean() - (np.dot(data, exp_x_b) / sum_exp_x_b) - b  # I'm not sure if np.dot is ever faster
x2 = -b * np.log(sum_exp_x_b / n) - a
```

File: scipy/stats/_continuous_distns.py
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:2994]: If we know what sort of data ("negative data", "very close values", ...) causes issues it would be an improvement to say that over "some data"

File: scipy/stats/tests/test_distributions.py
Comment (sethtroisi) [inline scipy/stats/tests/test_distributions.py:32]: you don't seem to make use of this import.

File: scipy/stats/tests/test_distributions.py
Comment (sethtroisi) [inline scipy/stats/tests/test_distributions.py:1058]: Wasn't this moved here in a previous PR or is that also pending?

**Review by sethtroisi — COMMENTED** [review]
Generally looks good, thanks for improving these!

File: scipy/stats/_continuous_distns.py
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:2984]: I verified that code before matches the parameter estimation given in source above.

File: scipy/stats/tests/test_distributions.py
Comment (swallan) [reply scipy/stats/tests/test_distributions.py:1058]: It is also pending in other PRs, that is correct.

**Comment by swallan** [comment]
@WarrenWeckesser, We were hoping that you could advise on an issue that is being encountered with this distribution. 

It is that if the the scale is close enough to zero for the input data, the system of equations runs into an overflow, causing warnings. Considering this situation seems to be identifiable by noticing that the standard deviation of the data is very small, @mdhaber and I took a look at the two simultaneous equations with the `limit b->0`, where `b` is the scale.

heuristically, it would seem that as `b` approaches zero, the location, `a` is the mean/mode of the data, as with such small scale all of the values are near identical.  But the limits are complicated and a bit out of my skill level. 

Screen Shot 2020-08-26 at 1 03 07 PM

Screen Shot 2020-08-26 at 1 02 31 PM


So, the real question is how to identify / and or treat this situation. In the least we could try to identify data with little to no deviations and pass them to the super method. (in addition to falling back on the super method if optimization over the system of equations fails). Another option may be to find solutions with the limits, if they are indeed computable. As you may also see, tests are failing because of attempts to suppress the warnings created during optimization, and a test exists to prohibit warning suppression. If there is an acceptable way to suppress warnings let me know. 

Example data that apply in this scenario would be: 

```

np.ones(1000)
Out[5]: 
array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., ...


gumbel_r.rvs(size=1000, loc=3, scale=1e-15)
Out[6]: 
array([3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., ...


gumbel_r.rvs(size=1000, loc=3, scale=1e-5)
Out[7]: 
array([3.00005101, 3.00000386, 3.00001233, 2.9999961 , 2.99999529,
       3.00002184, 3.00004064, 3.00001937, 3.00000083, 3.00004548,
       3.00001907, 2.99999804, 2.99999667, ... 
```
Where the optimization fails due to overflow.

**Comment by sethtroisi** [comment]
```
if x is a constant value
x0 = x[0]  # == x.mean()

b = x.mean() - sum(xi * exp(-xi/b)) / sum(exp(-xi/b))
b = x0 - sum(x0 * exp(-x0/b)) / sum(exp(-x0/b))
b = x0 - (x0 * exp(-x0/b)) / exp(-x0/b)
b = x0 - x0 
b = 0

a = -b * log(1/n * sum(exp(-x0 / b))
a = -b * log(1/n * n * exp(-x0/b))
a = -b * log(exp(-x0/b))
a = -b * -x0 / b
a = x0
```

**Comment by swallan** [comment]
The fit method for `gumbel_r` can be repurposed to for use with `gumbel_l`. The process is to use `gumbel_r.fit` with `data * -1` and then use the negative of the resulting location along with the unmodified scale: 

```

from scipy.stats import gumbel_r, gumbel_l
data = gumbel_l.rvs(size=1000, scale=2, loc=.55)

loc_r, scale_r = gumbel_r.fit(data * -1)
loc_l, scale_l = (- loc_r, scale_r)

loc_super, scale_super = gumbel_l.fit(data)  # gumbel_l is not overriden, so it uses rv_continuous. 


print(f"using system of equations: { scale_l:>5} \nusing super method:        { scale_super:>5}\n")

args = [data, (gumbel_l._fitstart(data),)]
func = gumbel_l._reduce_func(args, {})[1]

print("log-likelihood results:")
print(func((loc_l, scale_l), data))
print(func((loc_super, scale_super), data))


print(f"is system of equations less than super method?: {func((loc_l, scale_l), data) < func((loc_super, scale_super), data)}")

```

I believe this is a result of gumbel_r and gumbel_l being reflections of each other over their respective locations, observationally (charts are generated using code from [distribution documentation page](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.gumbel_r.html#scipy.stats.gumbel_r) + added title):
image
image

File: scipy/stats/tests/test_distributions.py
Comment (swallan) [reply scipy/stats/tests/test_distributions.py:1106]: Thanks for the reminder @sethtroisi - 

In this case, where I already merged scipy/master with this PR, what is the best method to go back and remove this addition from the commit it is in?

In a testing duplicate branch, it appears that when I go back and edit that commit in interactive rebase that it co-authors me on every commit including those that were added from the merge. 

So instead, I am thinking of
1. doing a hard reset to prior to the master branch merge
2. interactive rebase to amend the commit where this code was added
3. merge scipy/master back into the branch
4. force push to update 

Feels a bit cumbersome, so before I do any of that, is there a more efficient way to do it that you had in mind?

Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:1106]: >In this case, where I already merged scipy/master with this PR

If this means you locally merged scipy/master locally since gh-12457 merged, then you have two copies of `_assert_lessthan_loglike` in the file, right? If that's the case, just delete one, commit, and push normally.

Comment (swallan) [reply scipy/stats/tests/test_distributions.py:1106]: You're right – I was thinking about removing the commit it came from, but that is much simpler.

File: scipy/stats/_continuous_distns.py
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:3012]: There's a test error in `test_warning_calls_filters` which I think is related to this line.

See https://github.com/scipy/scipy/blob/master/scipy/_lib/tests/test_warnings.py#L84

```
[gw1] win32 -- Python 3.6.8 C:\hostedtoolcache\windows\Python\3.6.8\x64\python.exe
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\site-packages\scipy\_lib\tests\test_warnings.py:76: in warning_calls
    finder.visit(tree)
        bad_filters = ['optimize\\optimize.py:947', 'sparse\\__init__.py:245']
        bad_stacklevels = ['__init__.py:135', 'cluster\\hierarchy.py:3528', 'cluster\\vq.py:135', 'cluster\\vq.py:575', 'cluster\\vq.py:737', 'constants\\codata.py:1568', ...]
        base       = WindowsPath('C:/hostedtoolcache/windows/Python/3.6.8/x64/lib/site-packages/scipy')
        file       = <_io.TextIOWrapper name='C:\\hostedtoolcache\\windows\\Python\\3.6.8\\x64\\lib\\site-packages\\scipy\\stats\\_continuous_distns.py' mode='r' encoding='utf-8'>
        finder     = <scipy._lib.tests.test_warnings.FindFuncs object at 0x0000020F0773B550>
        path       = WindowsPath('C:/hostedtoolcache/windows/Python/3.6.8/x64/lib/site-packages/scipy/stats/_continuous_distns.py')
        tree       = <_ast.Module object at 0x0000020F07734FD0>
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\ast.py:253: in visit
    return visitor(node)
        method     = 'visit_Module'
        node       = <_ast.Module object at 0x0000020F07734FD0>
        self       = <scipy._lib.tests.test_warnings.FindFuncs object at 0x0000020F0773B550>
        visitor    = <bound method NodeVisitor.generic_visit of <scipy._lib.tests.test_warnings.FindFuncs object at 0x0000020F0773B550>>
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\ast.py:261: in generic_visit
```

**Review by sethtroisi — COMMENTED** [review]
I didn't verify benchmark and I left one comment about broken test, otherwise this seems ready to be committed.

**Comment by swallan** [comment]
@sethtroisi Thanks for taking a look - I know that the warning suppression is causing that failure and am still trying to figure out the best way to avoid it. 

Benchmarks are behaving as expected; Two free parameters is fitted quickly, fixing either one is still slow: 
(header is cut off in the screenshot, so here it is):
 distribution   |    loc_fixed     |     scale_fixed    |      shape1_fixed     |     True / True     |     True / False     |     False / True   False / False 
Screen Shot 2020-09-10 at 3 34 51 PM


I was hoping that @WarrenWeckesser would have an opinion on it; Whether it is better to try to perform the optimization and figure out some way to catch the warning (and then decide how to proceed if the optimization was successful) or to calculate in advance whether or not we expect there to be a warning by checking if there is low standard deviation.

**Comment by swallan** [comment]
The way I see it - there are essentially two roads to go down.
1. try to catch the type of scenarios that will causes warnings to come up
2. suppress the warnings during optimization in a way that does not fail tests that check for warning suppression

I lean towards finding a way to perform the optimization inside the overridden fit method, while suppressing warnings that may arise, and then checking for success of the optimization. It seems a bit unsure to me to try to determine this unless we can find the exact threshold that causes the overflow and how to test for it. Therefore, I am wondering if there is a way to get past `scipy._lib.tests.test_warnings.FindFuncs`.

**Comment by WarrenWeckesser** [comment]
I created a PR on swallan's repo with some suggested changes for this PR: https://github.com/swallan/scipy/pull/2

**Comment by WarrenWeckesser** [comment]
This needs to be updated to use `_assert_less_or_close_loglike` instead of `_assert_lessthan_loglike` (see https://github.com/scipy/scipy/pull/12882).  Maybe rebase on master and update the test?

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [inline scipy/stats/_continuous_distns.py:3095]: `data` comes directly from the user, and it might be a list, so we have to pass it through `asarray` before we can negate it with `-`.
```suggestion
        loc_r, scale_r, = gumbel_r.fit(-np.asarray(data), *args, **kwds)
```

**Review by WarrenWeckesser — APPROVED** [review]
There is one change needed, otherwise looks good.  I'll commit that change, and once the test suite completes again, merge the PR.

**Comment by WarrenWeckesser** [comment]
Looks like the `fit` method of the `rayleigh` distribution needs some work.  I've seen failures locally like those in CI, and it might just be luck that we haven't seen those failures in the master branch.

I think I know what the problem is; I'll work on a fix for `rayleigh.fit` today.

File: benchmarks/benchmarks/stats.py
Comment (swallan) [inline benchmarks/benchmarks/stats.py:192]:
```suggestion
    dists = ["pareto", "laplace", "rayleigh",
             "invgauss", "gumbel_r", "gumbel_l"
             "logistic"]
```

**Comment by WarrenWeckesser** [comment]
gh-12968 should fix the failures that we see in this PR.  If we get that PR (or something like it) merged, I expect this PR will run the test suite with no problems.

**Comment by WarrenWeckesser** [comment]
This should be OK to merge once the tests finish.

**Comment by mdhaber** [comment]
CircleCI finished successfully, even though it didn't report to GitHub properly. Travis failures are unrelated (arpack, minimum_phase)

**Comment by WarrenWeckesser** [comment]
Test failures were unrelated to this PR.

Merged; thanks @swallan and all reviewers.

**Comment by mdhaber** [comment]
We need to revisit this.

```python3
import numpy as np
from scipy import stats

np.random.seed(0)

dist = stats.gumbel_r
# dist = stats.gumbel_l
x = dist.rvs(loc=1, scale=2, size=100)

fparams = {'floc': 1}
afit = dist.fit(x, **fparams)  # subclass fit method
nfit = super(type(dist), dist).fit(x, **fparams)  # superclass fit method

print(dist.nnlf(nfit, x) - dist.nnlf(afit, x))  # 0
``` 
So far so good. As we expect, if one parameter is given, fall back to the superclass fit method.

When we do the same thing with `dist = stats.gumbel_l`:
```
print(dist.nnlf(nfit, x) - dist.nnlf(afit, x))  # -24.162130298986284
```
I notice that `afit` is `(-1, 3.01820809418259)` (fitted loc is *negative* 1) so it has something to do with the way `gumbel_l` is the reflection of `gumbel_r`.  

But also, we thought before that if one parameter is given, we couldn't rely on the optimal fit values given in the reference. That's not quite true.  If the `fscale` is given, we can still use the same formula for `loc`. If the `floc` is given, there is an equation for `scale`, although it's a little different than the one in the book when both are free. (I think that it simplifies when both are free? In any case, it's just the first-order condition corresponding with the scale.)

**Pull request merged**
